starky2/src/cross_table_lookup.rs

```diff
@@ -0,0 +1,294 @@
+use anyhow::{ensure, Result};
+use plonky2::field::extension_field::{Extendable, FieldExtension};
+use plonky2::field::field_types::Field;
+use plonky2::field::packed_field::PackedField;
+use plonky2::field::polynomial::PolynomialValues;
+use plonky2::hash::hash_types::RichField;
+use plonky2::iop::challenger::Challenger;
+use plonky2::plonk::config::GenericConfig;
+use plonky2::plonk::plonk_common::reduce_with_powers;
+use plonky2::util::reducing::ReducingFactor;
+
+use crate::all_stark::Table;
+use crate::config::StarkConfig;
+use crate::constraint_consumer::ConstraintConsumer;
+use crate::permutation::{
+    get_grand_product_challenge_set, GrandProductChallenge, GrandProductChallengeSet,
+};
+use crate::proof::StarkProofWithPublicInputs;
+use crate::stark::Stark;
+use crate::vars::StarkEvaluationVars;
+
+#[derive(Clone)]
+pub struct CrossTableLookup {
+    pub looking_table: Table,
+    pub looking_columns: Vec<usize>,
+    pub looked_table: Table,
+    pub looked_columns: Vec<usize>,
+}
+
+impl CrossTableLookup {
+    pub fn new(
+        looking_table: Table,
+        looking_columns: Vec<usize>,
+        looked_table: Table,
+        looked_columns: Vec<usize>,
+    ) -> Self {
+        assert_eq!(looking_columns.len(), looked_columns.len());
+        Self {
+            looking_table,
+            looking_columns,
+            looked_table,
+            looked_columns,
+        }
+    }
+}
+
+/// Lookup data for one table.
+#[derive(Clone)]
+pub struct CtlData<F: Field> {
```

Comment:
I didn't really get the meaning of "control" in this context, do you think PermutationData or LookupData can work, or since both are not 100% accurate (per the other comment), maybe GrandProductData?

Reply:
Oh sorry I should have made this explicit somewhere, CTL is the acronym of Cross-Table Lookup. I used this to make names less verbose, like `permutation_cross_table_lookup_zs` -> `permutation_ctl_zs`.